Add tag production env guard.

ivankruchkoff opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
Addresses issue #2054 

## Relevant technical choices
As we can't override the env on a per test basis due to https://github.com/WordPress/WordPress/blob/f5713b8d717b398efe7de40c869f77543fb2068e/wp-includes/load.php#L195

We instead define a function in the namespace to provide custom values for our tests.
https://github.com/google/site-kit-wp/pull/3734/files#diff-cd9bfad4b0a994c130af2d7fce75605bab9e19e99ccea0a3124f6ca73e69c4f4R58
<!-- Please describe your changes. -->

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

eugene-manuilov reviewed (changes requested):
@ivankruchkoff have you seen the [`uopz_set_static`](https://www.php.net/manual/en/function.uopz-set-static.php) function? Seems like we can use it to reset static variable inside of the `wp_get_environment_type` function, can't we? Can you try it on your end and if it works update your tests to use it?

ivankruchkoff commented:
> @ivankruchkoff have you seen the [`uopz_set_static`](https://www.php.net/manual/en/function.uopz-set-static.php) function? Seems like we can use it to reset static variable inside of the `wp_get_environment_type` function, can't we? Can you try it on your end and if it works update your tests to use it?

Tried it and added it to our workflow.

eugene-manuilov commented on tests/phpunit/integration/Core/Tags/Guards/Tag_Production_GuardTest.php:
```diff
@@ -0,0 +1,52 @@
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Tests\Core\Tags\Guards {
```
Let's remove curly brackets since we don't use them anywhere.
```suggestion
namespace Google\Site_Kit\Tests\Core\Tags\Guards;
```

ivankruchkoff replied:
https://github.com/google/site-kit-wp/pull/3734/commits/5d02b7481e6870acc3b4ba472ce463e116432196

eugene-manuilov commented on tests/phpunit/integration/Core/Tags/Guards/Tag_Production_GuardTest.php:
```diff
@@ -0,0 +1,52 @@
+			$tagproduction = new Tag_Production_Guard();
+			uopz_set_static( 'wp_get_environment_type', array( 'current_env' => 'development' ) );
+			$this->assertFalse( $tagproduction->can_activate() );
+			uopz_set_static( 'wp_get_environment_type', array( 'current_env' => 'production' ) );
```
Let's probably revert to original value instead of reverting to `production`:
```php
$env_type      = wp_get_environment_type();
$tagproduction = new Tag_Production_Guard();

uopz_set_static( 'wp_get_environment_type', array( 'current_env' => 'development' ) );
$this->assertFalse( $tagproduction->can_activate() );

uopz_set_static( 'wp_get_environment_type', array( 'current_env' => $env_type ) );
```

ivankruchkoff replied:
Updated in https://github.com/google/site-kit-wp/pull/3734/commits/ff826d110dfc2bc829d668209619d53a8ad05fe0 but the way the other tests are written, if the value is anything but production, they will fail.

eugene-manuilov reviewed (changes requested):
That's awesome! Thanks, @ivankruchkoff! Added two minor comments and it should be good to go.

eugene-manuilov commented on tests/phpunit/integration/Core/Tags/Guards/Tag_Production_GuardTest.php:
```diff
@@ -0,0 +1,51 @@
+
+class Tag_Production_GuardTest extends TestCase {
+
+	public function test_can_activate_on_older_environments() {
+		// Pre WP-5.5.0
+		if ( ! function_exists( 'wp_get_environment_type' ) ) {
+			$this->markTestSkipped( 'Missing wp_get_environment_type() function.' );
+		}
+		$tagproduction = new Tag_Production_Guard();
+		$this->assertTrue( $tagproduction->can_activate() );
+	}
```
I think if we want to test it for older versions of WP, we need to skip this test if the function exists, right?

eugene-manuilov replied:
Maybe let's rename the method name to be something like `test_can_activate_on_older_versions`?

ivankruchkoff replied:
https://github.com/google/site-kit-wp/pull/3734/commits/a81ec2efe6837400ef41a3b32d8f2d40b0dbb31a

ivankruchkoff replied:
Updated commenting for both in https://github.com/google/site-kit-wp/pull/3734/commits/267859182074ef4640ca2902182371aeb6596854

First test `test_can_activate_on_post_5_5_version()` is for post 5.5.0 scenarios, and thus skips if `wp_get_environment_type()` is not defined.
Second test `test_can_activate_on_older_versions()` is for pre 5.5.0 scenarios, and thus skips if `wp_get_environment_type()` is defined.

eugene-manuilov commented on tests/phpunit/integration/Core/Tags/Guards/Tag_Production_GuardTest.php:
```diff
@@ -0,0 +1,51 @@
+		$this->assertTrue( $tagproduction->can_activate() );
+	}
+
+	public function test_can_activate_in_production() {
+		// Post WP-5.5.0
+		if ( function_exists( 'wp_get_environment_type' ) ) {
+			$this->markTestSkipped( 'Testing legacy environments without wp_get_environment_type() function.' );
+		}
+		$tagproduction = new Tag_Production_Guard();
+		$this->assertTrue( $tagproduction->can_activate() );
+	}
```
Here is the opposite situation, we need to skip this test if the function doesn't exist which means that it is the old version of WP.

eugene-manuilov replied:
Same here, maybe let's rename the method name to be something like `test_can_activate_on_post_5_5_version`?

ivankruchkoff replied:
https://github.com/google/site-kit-wp/pull/3734/commits/a81ec2efe6837400ef41a3b32d8f2d40b0dbb31a

eugene-manuilov reviewed (changes requested):
Ok, thanks, @ivankruchkoff. The last comments from me and it should good to go.

eugene-manuilov commented:
@ivankruchkoff looks like you missed two previous comments: https://github.com/google/site-kit-wp/pull/3734#discussion_r675531996 and https://github.com/google/site-kit-wp/pull/3734#discussion_r675532732. Please, take a look.

ivankruchkoff commented:
> @ivankruchkoff looks like you missed two previous comments: [#3734 (comment)](https://github.com/google/site-kit-wp/pull/3734#discussion_r675531996) and [#3734 (comment)](https://github.com/google/site-kit-wp/pull/3734#discussion_r675532732). Please, take a look.

Addressed, the suggested function names were inverted, so I added clarifying comments.

eugene-manuilov reviewed (approved):
Thanks, @ivankruchkoff. This looks great to me!
